<div style="
    border: 2px solid #4CAF50; 
    padding: 15px; 
    background-color: #f4f4f4; 
    border-radius: 10px; 
    align-items: center;">

<h1 style="margin: 0; color: #4CAF50;">Neural Networks: Numpy, PyTorch, Google Colab</h1>
<h2 style="margin: 5px 0; color: #555;">DSAI</h2>
<h3 style="margin: 5px 0; color: #555;">Jakob Eggl</h3>

<div style="flex-shrink: 0;">
    <img src="https://www.htl-grieskirchen.at/wp/wp-content/uploads/2022/11/logo_bildschirm-1024x503.png" alt="Logo" style="width: 250px; height: auto;"/>
</div>
<p1> © 2025/26 Jakob Eggl. Nutzung oder Verbreitung nur mit ausdrücklicher Genehmigung des Autors.</p1>
</div>
<div style="flex: 1;">
</div>   

# Numpy

Wir wollen uns nun ``numpy`` widmen. Es ist eine grundlegende - wenn nicht sogar *DIE* - Bibliothek, auf der die meisten weiteren (und von uns auch bisher schon verwendeten) Bibliotheken aufbauen.

Für numpy gibt es viele Tutorials online, unter anderem wurden [Tutorial1](https://numpy.org/doc/stable/user/quickstart.html) und [Tutorial2](https://cs231n.github.io/python-numpy-tutorial/) für diesen Abschnitt des Notebooks verwendet.

## Was ist Numpy?

* **Num**erical **Py**thon
* Open Source
* ermöglicht numerisches Rechnen mit Python
* Hauptobjekte in Numpy: (mehrdimensionale) *Arrays*

Bisher haben wir uns manchmal schon implizit mit ``numpy`` Objekten beschäftigt:
* Bei scikit-learn (``sklearn``) werden haben wir oft mit numpy Objekten gearbeitet (zBsp.: in vielen Fällen waren *X* und *y* numpy arrays).
* Pandas Dataframes können auch einfach in numpy arrays konvertiert werden.

Bisher war das aber bei uns immer ein eher unwichtiges Detail. Wir betrachten nun weitere Gründe, warum wir uns gut in numpy auskennen sollen.

Wir importieren ``numpy`` ganz einfach mit:

In [2]:
import numpy as np

## Warum Numpy?

#### Vorteile:
* **sehr schnell** und **effizient** (ist in *C* geschrieben)
* Viele vektorisierte (vectorized) Operationen
* Große Anzahl an mathematischen Funktionen und Operationen implementiert
* Super integriert in alle gängigen Bibliotheken
* Effizienter Speicherverbrauch (überall gleicher Datentyp in einem array)

#### Nachteile:
* (Fast) nur numerische Datentypen (Integer, Float, Double, etc.) möglich
* Pro array nur ein Datentyp erlaubt (viele Datasets bei uns haben bisher oft auch andere Features gehabt (Name, Adresse, etc.))
* Namen der Features und Label(s) gehen verloren.

Betrachten wir den Geschwindigkeitsunterschied anhand von einem kurzem Beispiel:

In [3]:
import time

n_elements = 3000000

a = list(range(n_elements))
b = list(range(n_elements))

start = time.time()
z = zip(a, b)
c = [x + y for x, y in z]
time_1 = time.time() - start
print(f"Listenzeit: {time_1}")

a_np = np.array(a)
b_np = np.array(b)

start = time.time()
c_np = a_np + b_np
time_2 = time.time() - start
print(f"NumPy-Zeit: {time_2}")

print(f"Numpy ist in diesem Fall {time_1/time_2} mal schneller")

Listenzeit: 0.11680150032043457
NumPy-Zeit: 0.005227804183959961
Numpy ist in diesem Fall 22.342363296392577 mal schneller


![Vectored_Meme](../resources/Vectored.png)

(von https://imageresizer.com/pt/gerador-de-memes/editar/you-just-got-vectored)

## Numpy Arrays

Nun beschäftigen wir uns mit dem zentralen Element in numpy: Das **numpy array**.

In [4]:
# Short Recap on Python Lists

a_list = list(range(10))

print(a_list)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [5]:
a_list[1]=2
print(a_list)

[0, 2, 2, 3, 4, 5, 6, 7, 8, 9]


In [6]:
# Now to numpy arrays. We now convert our list to a numpy array.

a_array = np.array(a_list) # we only need to pass a list to the np.array function

In [7]:
print(a_array)
print(type(a_array))
print(a_array.dtype)

[0 2 2 3 4 5 6 7 8 9]
<class 'numpy.ndarray'>
int64


In [8]:
# Now what about this list?
a_second_list = ['hello', 'world!']

In [9]:
a_second_list

['hello', 'world!']

In [10]:
# Converting to numpy?
a_second_array = np.array(a_second_list)

In [11]:
print(a_second_array)
print(type(a_second_array))
print(a_second_array.dtype)

['hello' 'world!']
<class 'numpy.ndarray'>
<U6


Es funktioniert also schon, jedoch wird nun das ganze Array als *<U6* Datentyp gespeichert. Das sind unicode strings mit maximal 6 Zeichen.

Somit können wir auch die Texte "kürzen".

In [12]:
np.array(["hello", "world!"], dtype="U5")

array(['hello', 'world'], dtype='<U5')

In [13]:
# Extreme Example
np.array(["hello", "world!"], dtype="U2")

array(['he', 'wo'], dtype='<U2')

In [14]:
np.array(["hello", "world!"], dtype="U7")

array(['hello', 'world!'], dtype='<U7')

Es gibt noch weitere Informationen, welche wir über das Array herausfinden können.

In [15]:
# numpy arrays also have other properties that we can use to get information about the array
a = np.array([1, 2, 3, 4, 5])
print(a.shape)
print(a.ndim)
print(a.size)

(5,)
1
5


Somit ist *a* bei uns ein Vektor der Länge 5. Wir können aber auch eine Matrix erstellen.

In [16]:
# matrix in numpy
a = np.array([[1, 2, 3], [4, 5, 6]])

In [17]:
print(a)
print(a.shape)
print(a.ndim)
print(a.size)

[[1 2 3]
 [4 5 6]]
(2, 3)
2
6


Wir haben also nun eine Matrix mit 2 Zeilen (erster Eintrag in ``.shape``) und 3 Spalten (zweiter Eintrag). Dies ist genau gleich wie in der Mathematik.

**Merkregel:** Zeilen zuerst, Spalten später. 

### Operationen mit Numpy Arrays

Im folgenden sehen wir einige Beispiele von Befehlen, die in numpy verwendet werden können um Arrays zu erstellen, bearbeiten und abzufragen. Diese Liste ist nur ein kleiner Teil aller Möglichkeiten, die es in ``numpy`` gibt.

Mit (einem Teil von) den folgenden Befehlen werden wir im Rahmen des Bauens von neuronalen Netzwerken arbeiten, weswegen wir sie kurz besprechen.

In [18]:
# We can also change the shape of the numpy array
reshaped_a = a.reshape(3, 2)  # we now switch the number of columns and rows.

In [19]:
# There are of course many other ways on how to create a numpy array, without manually writing down each entry
zeros_array = np.zeros((3, 4))  
ones_array = np.ones((2, 5))    
random_array = np.random.rand(2, 3)  # uniform between 0 and 1
random_neg1_pos1_array = np.random.uniform(-1, 1, (2, 3))  
seven_array = np.full((2, 3), 7)
print("*"*50)
print("Zeros Array:")
print(zeros_array)
print("*"*50)
print("Ones Array:")
print(ones_array)
print("*"*50)
print("Random Array:")
print(random_array)
print("*"*50)
print("Random -1 to 1 Array:")
print(random_neg1_pos1_array)
print("*"*50)
print("Constant 7 array")
print(seven_array)

**************************************************
Zeros Array:
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
**************************************************
Ones Array:
[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]]
**************************************************
Random Array:
[[0.29255354 0.67094282 0.38152982]
 [0.07006982 0.21032955 0.97910274]]
**************************************************
Random -1 to 1 Array:
[[ 0.4538663   0.8185449   0.41463727]
 [ 0.33390644 -0.5025175  -0.55260822]]
**************************************************
Constant 7 array
[[7 7 7]
 [7 7 7]]


> **Übung:** Was wäre mit den bekannten Methoden also der schnellste Weg, um eine $3\times 3$ Matrix zu erstellen, welche aufsteigend die Zahlen 1-9 beinhaltet?

## Indexing von Numpy Arrays

Es gibt viele Möglichkeiten, wie in numpy ein Array indiziert werden kann. Im einfachsten Fall ist das gleich wie in den meisten gängigen Programmiersprachen. Es gibt aber auch Eigenheiten (zBsp slicing) für numpy.

In [20]:
# Create the following rank 2 array with shape (3, 4)
# [[ 1  2  3  4]
#  [ 5  6  7  8]
#  [ 9 10 11 12]]
a = np.array([[1,2,3,4], [5,6,7,8], [9,10,11,12]]) # this can be done more elegantly (see previous exercise)

# Use slicing to pull out the subarray consisting of the first 2 rows
# and columns 1 and 2; b is the following array of shape (2, 2):
# [[2 3]
#  [6 7]]
b = a[:2, 1:3]
print(b) # prints [[2 3], [6 7]]

# A slice of an array is a view into the same data, so modifying it
# will modify the original array.
print(a[0, 1])   # Prints "2"
b[0, 0] = 77     # b[0, 0] is the same piece of data as a[0, 1]
print(a[0, 1])   # Prints "77"

[[2 3]
 [6 7]]
2
77


Wir sehen also auch, dass wir durch slicing immer noch auf das gleiche Array referenzieren! Es können auch beide Indexmethoden gemischt werden:

In [21]:
# Create the following rank 2 array with shape (3, 4)
# [[ 1  2  3  4]
#  [ 5  6  7  8]
#  [ 9 10 11 12]]
a = np.array([[1,2,3,4], [5,6,7,8], [9,10,11,12]])

# Two ways of accessing the data in the middle row of the array.
# Mixing integer indexing with slices yields an array of lower rank,
# while using only slices yields an array of the same rank as the
# original array:
row_r1 = a[1, :]    # Rank 1 view of the second row of a
row_r2 = a[1:2, :]  # Rank 2 view of the second row of a
print(row_r1, row_r1.shape)  # Prints "[5 6 7 8] (4,)"
print(row_r2, row_r2.shape)  # Prints "[[5 6 7 8]] (1, 4)"

# We can make the same distinction when accessing columns of an array:
col_r1 = a[:, 1]
col_r2 = a[:, 1:2]
print(col_r1, col_r1.shape)  # Prints "[ 2  6 10] (3,)"
print(col_r2, col_r2.shape)  # Prints "[[ 2]
                             #          [ 6]
                             #          [10]] (3, 1)"

[5 6 7 8] (4,)
[[5 6 7 8]] (1, 4)
[ 2  6 10] (3,)
[[ 2]
 [ 6]
 [10]] (3, 1)


Wir können auch numpy arrays mit anderen numpy arrays (zbsp integer-array) indizieren:

In [22]:
import numpy as np

a = np.array([[1,2], [3, 4], [5, 6]])
print(a)

# An example of integer array indexing.
# The returned array will have shape (3,) and
print(a[[0, 1, 2], [0, 1, 0]])  # Prints "[1 4 5]"

# The above example of integer array indexing is equivalent to this:
print(np.array([a[0, 0], a[1, 1], a[2, 0]]))  # Prints "[1 4 5]"

# When using integer array indexing, you can reuse the same
# element from the source array:
print(a[[0, 0], [1, 1]])  # Prints "[2 2]"

# Equivalent to the previous integer array indexing example
print(np.array([a[0, 1], a[0, 1]]))  # Prints "[2 2]"

[[1 2]
 [3 4]
 [5 6]]
[1 4 5]
[1 4 5]
[2 2]
[2 2]


Dies kann auch verwendet werden um zBsp ein Element von jeder Zeile zu erhalten oder verändern

In [23]:
# Create a new array from which we will select elements
a = np.array([[1,2,3], [4,5,6], [7,8,9], [10, 11, 12]])

print(a)  # prints "array([[ 1,  2,  3],
          #                [ 4,  5,  6],
          #                [ 7,  8,  9],
          #                [10, 11, 12]])"

# Create an array of indices
b = np.array([0, 2, 0, 1])

# Select one element from each row of a using the indices in b
print(a[np.arange(4), b])  # Prints "[ 1  6  7 11]"

# Mutate one element from each row of a using the indices in b
a[np.arange(4), b] += 10

print(a)  # prints "array([[11,  2,  3],
          #                [ 4,  5, 16],
          #                [17,  8,  9],
          #                [10, 21, 12]])

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
[ 1  6  7 11]
[[11  2  3]
 [ 4  5 16]
 [17  8  9]
 [10 21 12]]


Auch mit einem Boolean-Array kann indiziert werden.

In [24]:
import numpy as np

a = np.array([[1,2], [3, 4], [5, 6]])
print(a)

bool_idx = (a > 2)   # Find the elements of a that are bigger than 2;
                     # this returns a numpy array of Booleans of the same
                     # shape as a, where each slot of bool_idx tells
                     # whether that element of a is > 2.

print(bool_idx)      # Prints "[[False False]
                     #          [ True  True]
                     #          [ True  True]]"

# We use boolean array indexing to construct a rank 1 array
# consisting of the elements of a corresponding to the True values
# of bool_idx
print(a[bool_idx])  # Prints "[3 4 5 6]"

# We can do all of the above in a single concise statement:
print(a[a > 2])     # Prints "[3 4 5 6]"

[[1 2]
 [3 4]
 [5 6]]
[[False False]
 [ True  True]
 [ True  True]]
[3 4 5 6]
[3 4 5 6]


## Mathematische Funktionen und Matrizen/Vektorrechnung (Lineare Algebra)

Nachdem ja numpy vorrangig für die Durchführung von numerischen Simulationen/Programmen entwickelt wurde, möchten wir uns jetzt noch spannende und vor allem für den Machine Learning Bereich wichtige Funktionen/Operationen ansehen.

#### Addition und Subtraktion von Arrays

In [25]:
a = np.array([1,2,3])
b = np.array([4,5,6])

print(a+b)
print(a-b)
print(a+3*b)

[5 7 9]
[-3 -3 -3]
[13 17 21]


#### Transponieren

In [26]:
A = np.array([[1, 2], [3, 4], [5, 6]])
print(A)

[[1 2]
 [3 4]
 [5 6]]


In [27]:
# We can transpose the array using the .T attribute
transposed_A = A.T

In [28]:
print(transposed_A)

[[1 3 5]
 [2 4 6]]


#### Matrix-Matrix und Matrix-Vektor und Skalarprodukt und Betrag vom Vektor

In [29]:
A = np.array([[1, 2], [3, 4], [5, 6]])
B = np.array([[7, 8], [9, 10], [11, 12]])

# Matrix multiplication (dot product) using numpy
matrix_product = np.dot(A, B.T)
matrix_product_2 = A @ B.T

print(matrix_product)
print(matrix_product_2)

[[ 23  29  35]
 [ 53  67  81]
 [ 83 105 127]]
[[ 23  29  35]
 [ 53  67  81]
 [ 83 105 127]]


In [30]:
A = np.array([[1, 2, 3], [4, 5, 6]])
x = np.array([4,5,6])

print(A@x) # calculates A*x

[32 77]


In [31]:
# Scalarproduct (Innerproduct)
a = np.array([5,6,7])
b = np.array([0,1,1])
print(np.inner(a,b))

13


In [32]:
# Length of a vector
a = np.array([5,6,7])
print(np.linalg.norm(a)) # Calculates with Pythagorean theorem

10.488088481701515


#### Lösen von Gleichungssystemen

In [33]:
# We can also solve a system of linear equations using numpy
A = np.array([[3, 2], [1, 4]])
b = np.array([5, 6])
x = np.linalg.solve(A, b)  # Solves the equation Ax = b
print("Solution x:", x)  # Prints the solution to the system of equations

Solution x: [0.8 1.3]


> **Übung:** Wie sieht das Gleichungssystem hier aus?

In [34]:
# Note that this is not possible if the determinant of A is zero, e.g.
A = np.array([[1, 2], [2, 4]])
b = np.array([3, 6])
try:
    x = np.linalg.solve(A, b)  # This will raise an error
except:
    print("The matrix A is singular, cannot solve the system of equations.")

The matrix A is singular, cannot solve the system of equations.


In [35]:
# We can check that directly:
det_A = np.linalg.det(A)
print("Determinant of A:", det_A)  # Prints the determinant of A

Determinant of A: 0.0


In [36]:
# And we can also calculate the inverse of a matrix, if it is not singular
A = np.array([[1, 2], [3, 4]])
A_inv = np.linalg.inv(A)  # Calculates the inverse of A
print("Inverse of A:\n", A_inv)  # Prints the inverse of A

# Sanity check:
print(A@A_inv)

Inverse of A:
 [[-2.   1. ]
 [ 1.5 -0.5]]
[[1.00000000e+00 1.11022302e-16]
 [0.00000000e+00 1.00000000e+00]]


#### (Arg)Min und (Arg)Max

In [37]:
# We can also look for the maximum and minimum values in an array and its indices
a = np.array([1, 2, 3, 4, 5])
print(a)
max_value = np.max(a)  # Maximum value
max_index = np.argmax(a)  # Index of maximum value
min_value = np.min(a)  # Minimum value
min_index = np.argmin(a)  # Index of minimum value
print("Max value:", max_value, "at index", max_index)
print("Min value:", min_value, "at index", min_index)

[1 2 3 4 5]
Max value: 5 at index 4
Min value: 1 at index 0


In [38]:
# This can also be done in matrices
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(A)
max_value = np.max(A)  # Maximum value in the matrix
max_index = np.argmax(A)  # Index of maximum value in the matrix
min_value = np.min(A)  # Minimum value in the matrix
min_index = np.argmin(A)  # Index of minimum value in the matrix
print("Max value in matrix:", max_value, "at index", max_index)
print("Min value in matrix:", min_value, "at index", min_index)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
Max value in matrix: 9 at index 8
Min value in matrix: 1 at index 0


In [39]:
# And we can also get the maximum and minimum values and indices in each row or column

print(A)

max_values_row = np.max(A, axis=1)  # Maximum values in each row
print("Max values in each row:", max_values_row)
min_values_row = np.min(A, axis=1)  # Minimum values in each row
print("Min values in each row:", min_values_row)
# We can also get the maximum and minimum values in each column
max_values_col = np.max(A, axis=0)  # Maximum values in each column
print("Max values in each column:", max_values_col)


[[1 2 3]
 [4 5 6]
 [7 8 9]]
Max values in each row: [3 6 9]
Min values in each row: [1 4 7]
Max values in each column: [7 8 9]


#### Mathematische Funktionen und komplexe Zahlen

In [40]:
# We can also just apply functions like sin, abs, etc. to numpy arrays
a = np.array([1, 2, 3, -4, -5])
print("Original array:", a)
print("Sine of array:", np.sin(a))  # Applies sine function to each element
print("Absolute values of array:", np.abs(a))  # Applies absolute value function to each
print("Square root of array:", np.sqrt(np.abs(a)))  # Applies square root function to each element (after taking absolute value)

# We can also apply functions to matrices
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(A)
print("Sine of matrix:\n", np.sin(A))  # Applies sine function to each element
print("Absolute values of matrix:\n", np.abs(A))  # Applies absolute value function to each element


# And we can also work with complex numbers in numpy
complex_array = np.array([1 + 2j, 3 + 4j, 5 + 6j])
print("Complex array:", complex_array)
print("Real part of complex array:", np.real(complex_array))  # Extracts real part
print("Imaginary part of complex array:", np.imag(complex_array))  # Extracts

Original array: [ 1  2  3 -4 -5]
Sine of array: [0.84147098 0.90929743 0.14112001 0.7568025  0.95892427]
Absolute values of array: [1 2 3 4 5]
Square root of array: [1.         1.41421356 1.73205081 2.         2.23606798]
[[1 2 3]
 [4 5 6]
 [7 8 9]]
Sine of matrix:
 [[ 0.84147098  0.90929743  0.14112001]
 [-0.7568025  -0.95892427 -0.2794155 ]
 [ 0.6569866   0.98935825  0.41211849]]
Absolute values of matrix:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Complex array: [1.+2.j 3.+4.j 5.+6.j]
Real part of complex array: [1. 3. 5.]
Imaginary part of complex array: [2. 4. 6.]


# PyTorch

Nun beschäftigen wir uns mit ``PyTorch``. Es baut auf ``numpy`` und ist eine sehr weit verbreitete Bibliothek, die das Entwickeln von neuronalen Netzwerken sehr einfach ermöglicht.

Das folgende basiert hauptsächlich auf einem sehr gutem Tutorium von *Philipp Lippe* ([link](https://github.com/phlippe/uvadlc_notebooks/blob/master/docs/tutorial_notebooks/tutorial2/Introduction_to_PyTorch.ipynb)).

Wir starten mit dem Import von dem Paket, genannt ``torch``.

In [41]:
import torch
print("Using torch", torch.__version__)

Using torch 2.8.0+cpu


Sollte die obige Zelle nicht ausführbar sein, so müssen wir PyTorch zuerst installieren. Einen Command dazu kann man sich [hier](https://pytorch.org/get-started/locally/) erstellen lassen. Wichtig ist die GPU (Cuda) Funktionalität, welche man unbedingt auswählen sollte, sofern man eine (unterstützte) NVIDIA Grafikkarte besitzt.

**Hinweis:** Fast alle gängigen NVIDIA Grafikkarten werden unterstützt. Es wird auch Apple-Silicon unterstützt.

Nun kommen wir auch schon zu den wichtigsten Elementen in PyTorch (und somit quasi auch für unsere Neuronalen Netzwerke später), den ***Tensoren***.

#### Tensoren

* sind eine Generalisierung von Matrizen bzw. Vektoren (also mathematische Objekte)
* prinzipiell komplizierte Objekte, jedoch für unsere Zwecke mit Kenntnis von Matrizen und Vektoren recht leicht zu bedienen.
* namensgebend auch zum Beispiel für die Bibliothek ``Tensorflow`` (hat den gleichen Zweck wie zBsp PyTorch)

![Tensor](../resources/Tensor_1.png)

(von https://databasecamp.de/python/tensor)

Wie könnte der ganz rechte Tensor vom obigen Bild aussehen in der Praxis?

![Columbus_Tensored](../resources/Columbus_Tensor.png)

Eine andere Ansicht zeigt das nochmal allgemeiner (und von einer anderen Sichtweise im Vergleich zu vorher)

![Tensor_2](../resources/Tensor_2.png)

(von https://brainpy.readthedocs.io/en/brainpy-1.1.x/tutorial_math/tensors.html)

Wir siehen hier auch die verschiedenen Dimensionen, genannt ``axis``. In welcher Reihenfolge die axis numeriert werden, hängt von der *shape* ab. 

So ist im obigen Bild beim "3D tensor" *axis 0* jene axis, die zum ersten Eintrag der Shape *(4, 3, 2)* gehört.

**Wichtig:** Wir können auch in NumPy mit *Tensoren* arbeiten, jedoch werden sie dort nicht Tensoren genannt, sondern nach wie vor einfach *arrays*. Sprich auch in Numpy können wir 3d arrays (können sogar $n$-dimensional sein) arbeiten!



Man muss sich in beiden Fällen (NumPy und PyTorch) aber immer bewusst sein, welche Dimension zu welcher Richtung gehört. Das folgende Code-Beispiel sollte das verdeutlichen. Wir verwenden nochmal kurz NumPy am Anfang, danach sehen wir uns das ganze in PyTorch an.

> **Übung:** Finde Beispiele für 3D und 4D Tensoren im Bereich Machine Learning. **Tipp**: Welche Art von Tensor waren unsere bisherigen "tabellarischen Daten" immer?

In [42]:
import scipy
import sklearn
print(scipy.__version__)
print(sklearn.__version__)

1.13.1
1.6.1


In [43]:
import sklearn.datasets as sklearn_datasets
import numpy as np

In [44]:
iris_data = sklearn_datasets.load_iris()

In [45]:
iris_data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

Wir sind interessiert am ``.data`` Attribut, da es das eigentliche Array beinhaltet.

In [46]:
X = iris_data.data

In [47]:
print(X)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

In [48]:
print(X.shape)
print(type(X))
print(X.ndim)

(150, 4)
<class 'numpy.ndarray'>
2


Welche Ordnung hat also dieses Array (bzw. dieser Tensor)?

Nun hat die erste Axis (axis 0), also 150 Einträge und besteht also aus unseren verschiedenen Datenpunkten. Wobei jeder Datenpunkt ein Vektor von 4 Einträgen ist. Also haben wir 4 Features. Es ist also in unserem Fall die erste Axis (axis 0) die Datenpunkt-Axis und die zweite Axis (axis 1) die Feature axis.

In [49]:
# What happens when we run this code?
X_aggregated = np.mean(X, axis=0)

In [50]:
# How large do you think is now the resulting array "X_aggregated"?
print(X_aggregated)
print(X_aggregated.shape)

[5.84333333 3.05733333 3.758      1.19933333]
(4,)


Was wäre, wenn wir ``axis=1`` verwendet hätten?

In [51]:
X_aggregated_2 = np.mean(X, axis=1)

In [52]:
# How large do you think is now the resulting array "X_aggregated"?
print(X_aggregated_2)
print(X_aggregated_2.shape)

[2.55  2.375 2.35  2.35  2.55  2.85  2.425 2.525 2.225 2.4   2.7   2.5
 2.325 2.125 2.8   3.    2.75  2.575 2.875 2.675 2.675 2.675 2.35  2.65
 2.575 2.45  2.6   2.6   2.55  2.425 2.425 2.675 2.725 2.825 2.425 2.4
 2.625 2.5   2.225 2.55  2.525 2.1   2.275 2.675 2.8   2.375 2.675 2.35
 2.675 2.475 4.075 3.9   4.1   3.275 3.85  3.575 3.975 2.9   3.85  3.3
 2.875 3.65  3.3   3.775 3.35  3.9   3.65  3.4   3.6   3.275 3.925 3.55
 3.8   3.7   3.725 3.85  3.95  4.1   3.725 3.2   3.2   3.15  3.4   3.85
 3.6   3.875 4.    3.575 3.5   3.325 3.425 3.775 3.4   2.9   3.45  3.525
 3.525 3.675 2.925 3.475 4.525 3.875 4.525 4.15  4.375 4.825 3.4   4.575
 4.2   4.85  4.2   4.075 4.35  3.8   4.025 4.3   4.2   5.1   4.875 3.675
 4.525 3.825 4.8   3.925 4.45  4.55  3.9   3.95  4.225 4.4   4.55  5.025
 4.25  3.925 3.925 4.775 4.425 4.2   3.9   4.375 4.45  4.35  3.875 4.55
 4.55  4.3   3.925 4.175 4.325 3.95 ]
(150,)


Und natürlich können wir auch die Richtung umdrehen. Sprich in unserem Fall die Matrix $X$ transponieren.

**Hinweis:** Das Transponieren von Tensoren funktioniert auch in höheren Dimensionen. Hier muss man jedoch eine Reihenfolge angeben, in der die verschiedenen *axis* getauscht werden sollen.

In [53]:
flipped_X = X.T 
print(flipped_X)
print(flipped_X.shape)

[[5.1 4.9 4.7 4.6 5.  5.4 4.6 5.  4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
  5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.  5.  5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.
  5.5 4.9 4.4 5.1 5.  4.5 4.4 5.  5.1 4.8 5.1 4.6 5.3 5.  7.  6.4 6.9 5.5
  6.5 5.7 6.3 4.9 6.6 5.2 5.  5.9 6.  6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
  6.3 6.1 6.4 6.6 6.8 6.7 6.  5.7 5.5 5.5 5.8 6.  5.4 6.  6.7 6.3 5.6 5.5
  5.5 6.1 5.8 5.  5.6 5.7 5.7 6.2 5.1 5.7 6.3 5.8 7.1 6.3 6.5 7.6 4.9 7.3
  6.7 7.2 6.5 6.4 6.8 5.7 5.8 6.4 6.5 7.7 7.7 6.  6.9 5.6 7.7 6.3 6.7 7.2
  6.2 6.1 6.4 7.2 7.4 7.9 6.4 6.3 6.1 7.7 6.3 6.4 6.  6.9 6.7 6.9 5.8 6.8
  6.7 6.7 6.3 6.5 6.2 5.9]
 [3.5 3.  3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 3.7 3.4 3.  3.  4.  4.4 3.9 3.5
  3.8 3.8 3.4 3.7 3.6 3.3 3.4 3.  3.4 3.5 3.4 3.2 3.1 3.4 4.1 4.2 3.1 3.2
  3.5 3.6 3.  3.4 3.5 2.3 3.2 3.5 3.8 3.  3.8 3.2 3.7 3.3 3.2 3.2 3.1 2.3
  2.8 2.8 3.3 2.4 2.9 2.7 2.  3.  2.2 2.9 2.9 3.1 3.  2.7 2.2 2.5 3.2 2.8
  2.5 2.8 2.9 3.  2.8 3.  2.9 2.6 2.4 2.4 2.7 2.7 3.  3.4 3.1 2.3 3.  2.5
  2.6 3.  2.

Was ist nun die axis0 und was ist die axis1?

Wir können aber auch ohne angegebener Axis die Daten aggregieren. Was ist das Ergebnis von der folgenden Code-Zeile?

In [54]:
X_aggregated_3 = np.mean(X)

In [55]:
print(X_aggregated_3)
print(X_aggregated_3.shape)

3.4644999999999997
()


Wie sieht das ganze in PyTorch aus?

In [56]:
import torch

In [57]:
X = iris_data.data

In [58]:
# Now we convert the numpy array to a PyTorch tensor
X_tensor = torch.tensor(X, dtype=torch.float32) # We can also skip the dtype, then it is inferred from the numpy array but it is a good practice to specify the dtype

# Another possibility is
X_tensor_2 = torch.from_numpy(X)

In [59]:
print(X_tensor)

tensor([[5.1000, 3.5000, 1.4000, 0.2000],
        [4.9000, 3.0000, 1.4000, 0.2000],
        [4.7000, 3.2000, 1.3000, 0.2000],
        [4.6000, 3.1000, 1.5000, 0.2000],
        [5.0000, 3.6000, 1.4000, 0.2000],
        [5.4000, 3.9000, 1.7000, 0.4000],
        [4.6000, 3.4000, 1.4000, 0.3000],
        [5.0000, 3.4000, 1.5000, 0.2000],
        [4.4000, 2.9000, 1.4000, 0.2000],
        [4.9000, 3.1000, 1.5000, 0.1000],
        [5.4000, 3.7000, 1.5000, 0.2000],
        [4.8000, 3.4000, 1.6000, 0.2000],
        [4.8000, 3.0000, 1.4000, 0.1000],
        [4.3000, 3.0000, 1.1000, 0.1000],
        [5.8000, 4.0000, 1.2000, 0.2000],
        [5.7000, 4.4000, 1.5000, 0.4000],
        [5.4000, 3.9000, 1.3000, 0.4000],
        [5.1000, 3.5000, 1.4000, 0.3000],
        [5.7000, 3.8000, 1.7000, 0.3000],
        [5.1000, 3.8000, 1.5000, 0.3000],
        [5.4000, 3.4000, 1.7000, 0.2000],
        [5.1000, 3.7000, 1.5000, 0.4000],
        [4.6000, 3.6000, 1.0000, 0.2000],
        [5.1000, 3.3000, 1.7000, 0

In [60]:
print(X_tensor_2)

tensor([[5.1000, 3.5000, 1.4000, 0.2000],
        [4.9000, 3.0000, 1.4000, 0.2000],
        [4.7000, 3.2000, 1.3000, 0.2000],
        [4.6000, 3.1000, 1.5000, 0.2000],
        [5.0000, 3.6000, 1.4000, 0.2000],
        [5.4000, 3.9000, 1.7000, 0.4000],
        [4.6000, 3.4000, 1.4000, 0.3000],
        [5.0000, 3.4000, 1.5000, 0.2000],
        [4.4000, 2.9000, 1.4000, 0.2000],
        [4.9000, 3.1000, 1.5000, 0.1000],
        [5.4000, 3.7000, 1.5000, 0.2000],
        [4.8000, 3.4000, 1.6000, 0.2000],
        [4.8000, 3.0000, 1.4000, 0.1000],
        [4.3000, 3.0000, 1.1000, 0.1000],
        [5.8000, 4.0000, 1.2000, 0.2000],
        [5.7000, 4.4000, 1.5000, 0.4000],
        [5.4000, 3.9000, 1.3000, 0.4000],
        [5.1000, 3.5000, 1.4000, 0.3000],
        [5.7000, 3.8000, 1.7000, 0.3000],
        [5.1000, 3.8000, 1.5000, 0.3000],
        [5.4000, 3.4000, 1.7000, 0.2000],
        [5.1000, 3.7000, 1.5000, 0.4000],
        [4.6000, 3.6000, 1.0000, 0.2000],
        [5.1000, 3.3000, 1.7000, 0

In [61]:
# Now we can do the same things as before
X_tensor_agg = torch.mean(X_tensor, dim=0)

In [62]:
print(X_tensor_agg)

tensor([5.8433, 3.0573, 3.7580, 1.1993])


Ein kleiner Unterschied ist, dass nun die Dimension wirklich mit ``dim`` abgekürzt wird und nicht mit ``axis``.

Wie kann man nun einen Tensor erstellen (ohne Daten)?

-> (Fast) genauso wie in numpy!

In [63]:
zeros_tensor = torch.zeros((3,5))
ones_tensor = torch.ones((2,5))
my_list = [1,2,3.0, 1.123345, -3]
my_tensor = torch.tensor(my_list)

In [64]:
print(zeros_tensor)
print(ones_tensor)
print(my_tensor)

tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
tensor([ 1.0000,  2.0000,  3.0000,  1.1233, -3.0000])


Und natürlich geht das auch mit mehreren Dimensionen (nicht nur 1 oder 2).

In [65]:
# Create 7d Tensor:
seven_d_tensor = torch.zeros((2, 3, 2, 2, 2, 3, 2))  # Example of a 7-dimensional tensor
print(seven_d_tensor)

tensor([[[[[[[0., 0.],
             [0., 0.],
             [0., 0.]],

            [[0., 0.],
             [0., 0.],
             [0., 0.]]],


           [[[0., 0.],
             [0., 0.],
             [0., 0.]],

            [[0., 0.],
             [0., 0.],
             [0., 0.]]]],



          [[[[0., 0.],
             [0., 0.],
             [0., 0.]],

            [[0., 0.],
             [0., 0.],
             [0., 0.]]],


           [[[0., 0.],
             [0., 0.],
             [0., 0.]],

            [[0., 0.],
             [0., 0.],
             [0., 0.]]]]],




         [[[[[0., 0.],
             [0., 0.],
             [0., 0.]],

            [[0., 0.],
             [0., 0.],
             [0., 0.]]],


           [[[0., 0.],
             [0., 0.],
             [0., 0.]],

            [[0., 0.],
             [0., 0.],
             [0., 0.]]]],



          [[[[0., 0.],
             [0., 0.],
             [0., 0.]],

            [[0., 0.],
             [0., 0.],
           

Bei solchen Dimensionen macht es dann natürlich wenig Sinn, sich die Daten direkt anzusehen, aber man kann sich natürlich die shape ansehen.

In [66]:
# Shape of 7d tensor
print(seven_d_tensor.shape)

torch.Size([2, 3, 2, 2, 2, 3, 2])


> **Übung:** Welche Shape hat ein RGB FullHD Film, welcher in 24 FPS abgespielt wird und eine Länge von 90min hat? Wie viele Einträge hat der Tensor? Welche Reihenfolge ist sinnvoll?

> **Übung:** Was wäre, wenn wir nun 12 so Videos gleichzeitig in einem Tensor speichern möchten? Können die Videos eine unterschiedliche Länge haben?

Ok, aber warum nutzen wir jetzt genau PyTorch und nicht einfach numpy, wenn in beiden Bibliotheken Matrizen erstellt werden können?

Grund dafür sind (mitunter) folgende 3 Dinge:
* Die wichtigsten Machine Learning Funktionen sind bereits implementiert (werden wir noch lernen)
* Automatische Differenzierung, welche das Trainieren von neuronalen Netzwerken ermöglicht
* Die sehr einfache Möglichkeit, die Berechnungen auf einem Hardware-beschleunigten Gerät (zBsp. die GPU) auszuführen

#### Dynamic Computation Graph und Automatic Differentiation

Es ist also möglich in PyTorch, sich komplizierte Funktionen zu basteln (zum Beispiel ein Neuronales Netz) und PyTorch kann automatisch die Ableitungen bzgl. den verschiedenen Parametern berechnen.

Warum das Berechnen der Ableitungen wichtig ist, werden wir in den nächsten Notebooks noch genauer erfahren (Gradient Descent). Kurz gesagt, gibt uns die Ableitung der Fehlerfunktion eine Richtung vor, in die wir uns bewegen müssen.

Sehen wir uns einmal an, wie wir auf die Gradienten zugreifen können.

In [67]:
x = torch.ones((3))
print(x)
print(x.requires_grad)

tensor([1., 1., 1.])
False


In [68]:
# We can also equip the tensor with the ability to calculate gradients
x.requires_grad_(True)
print(x.requires_grad)

True


Um nun eine Ableitung einer Funktion zu berechnen, definieren wir uns einmal eine beliebige Funktion, in diesem Fall:
$$y(x) = \frac{1}{n}\sum_{i=1}^{n} \left[(x_i + 2)^2 + 3\right],$$

wobei hier $n$ die Anzahl der Elemente von $x$ darstellt.

In unserem Fall ist nun $x$ der Parameter, den wir optimieren (also verändern) wollen, sodass $y(x)$ maximal (oder minimal) wird, sprich ein Extremum erreicht.

**Wichtig:** Später wird $y(x)$ die *Loss*-Funktion sein, und diese wollen wir natürlich minimieren. Somit kann man sich auch jetzt schon vorstellen, dass $y(x)$ eine Loss-Funktion ist.

Wir wissen, dass bei einem Extremum die Ableitung $0$ ist, somit müssen wir die Ableitung von $y(x)$ bezüglich $x$ finden, sprich $\frac{\mathrm d\, y(x)}{\mathrm d\, x}$.

Dabei kann uns jetzt PyTorch helfen.

Die Art, wie PyTorch die Ableitung berechnet ist mit sogenannten *Computational Graphs*. Um zu verstehen, was das ist, zerlegen wir nun unsere Funktion in kleinere Bestandteile.

Als Input verwenden wir den Vektor $x=[1,2,3]$ (zum Beispiel eine Person mit $1$ Geschwister, $2$ Urlauben pro Jahr und $3$ Mahlzeiten pro Tag).

In [69]:
x = torch.arange(3, dtype=torch.float32, requires_grad=True)  # Create a tensor with requires_grad set to True (we want to calculate the gradient later)
print(x)

tensor([0., 1., 2.], requires_grad=True)


In [70]:
a = x + 2
b = a ** 2
c = b + 3
y = c.mean()
print(f'y = {y}')
print("Y", y)

y = 12.666666984558105
Y tensor(12.6667, grad_fn=<MeanBackward0>)


Hier ist der *Computational Graph* nun auch grafisch dargestellt.

![Computational_Graph](../resources/pytorch_computation_graph.svg)

(von https://github.com/phlippe/uvadlc_notebooks/blob/master/docs/tutorial_notebooks/tutorial2/Introduction_to_PyTorch.ipynb)

Nun berechnen wir die Ableitung $\frac{\mathrm d\, y(x)}{\mathrm d\, x}$. Diese setzt sich (aufgrund der Kettenregel) aus den folgenden Ableitungen zusammen:

$$\frac{\mathrm d\, y(x)}{\mathrm d\, x_i} = \frac{\mathrm d\, y}{\mathrm d\, c_i}\cdot\frac{\mathrm d\, c_i}{\mathrm d\, b_i}\cdot \frac{\mathrm d\, b_i}{\mathrm d\, a_i}\cdot \frac{\mathrm d\, a_i}{\mathrm d\, x_i}.$$

**Hinweis:** Nachdem hier $x$ ein Vektor ist, müssen wir die Ableitung für jede Komponente einzeln berechnen. Dementsprechend wurde oben der Index $i$ verwendet.

Wie können wir die Ableitung händisch berechnen?

$$
\frac{\mathrm d\, a_i}{\mathrm d\, x_i} = 1,\hspace{1cm}
\frac{\mathrm d\, b_i}{\mathrm d\, a_i} = 2\cdot a_i,\hspace{1cm}
\frac{\mathrm d\, c_i}{\mathrm d\, b_i} = 1,\hspace{1cm}
\frac{\mathrm d\, y}{\mathrm d\, c_i} = \frac{1}{3}.
$$

Somit ergibt sich für $x=[0,1,2]$ und $a_i=x_i+2$ die Ableitung

$$\frac{\mathrm d\, y(x)}{\mathrm d\, x} =\left[\frac43,2,\frac83\right].$$

Wie kann das nun PyTorch machen?

In [71]:
# For calculating the gradient, we need to call the backward() method on the output tensor
y.backward()

In [72]:
# Then we get the gradient of y with respect to x by invoking x.grad
print("Gradient of y with respect to x:", x.grad)

Gradient of y with respect to x: tensor([1.3333, 2.0000, 2.6667])


**Wichtig:** Später werden wir die Ableitungen nicht bezüglich $x$ berechnen (wir können ja unsere Daten nicht ändern), sondern werden die Parameter bezüglich der Parameter von unserem Netzwerk berechnen. Diese können wir dann jeden Schritt dementsprechend anpassen.

**Hinweis:** Das heißt, bei einer Loss-Funktion sind die Daten fix und wir haben die Parameter der Funktion/des neuronalen Netzwerks als Parameter.

**Wichtig:** Nachdem hier unser Input ein Vektor ist, sprechen wir eigentlich nicht von einer Ableitung, sondern von einem **Gradient**. Ein Gradient ist ein Vektor, bei dem jede Komponente, die Ableitung bzgl. der Komponente ist.

#### GPU Support

Wie bereits oben angekündigt, wollen wir hier nun die GPU (leider nur NVIDIA (und Apple Silicon) möglich) benutzen, sofern eine zur Verfügung steht. Mit ihr können wir die Berechnungen dann mit *Cuda* durchführen. 

Ob eine GPU zur Verfügung steht, können wir mit folgenden Command testen.

**Hinweis:** Es kann nur dann die NVIDIA GPU verwendet werden, wenn auch die entsprechende PyTorch Version installiert ist.

**Hinweis:** Für alle ohne NVIDIA GPU gibt es auch einige Online Tools bei denen man gratis hardwarebeschleunigt Notebooks ausführen kann (siehe Ende vom Notebook).

In [73]:
gpu_avail = torch.cuda.is_available()
print(f'GPU is available: {gpu_avail}')

GPU is available: False


Im Allgemeinen ist es gängig, folgenden command am Beginn jedes Notebooks zu verwenden.

In [74]:
device = torch.device('cuda') if torch.cuda.is_available() else torch.device('mps') if torch.backends.mps.is_available() else torch.device('cpu')
print(device)

cpu


Um nun auch wirklich die Berechnungen auf der GPU durchzuführen, müssen wir unsere Berechnungen auf dieses Gerät schieben. Dies geht mit dem `to(device)` command.

In [75]:
x = torch.ones((3,2))
x = x.to(device)  # Move the tensor to the GPU if available
print(x)

tensor([[1., 1.],
        [1., 1.],
        [1., 1.]])


In [76]:
# We can also directly create a tensor on the GPU
x_gpu = torch.ones((3, 2), device=device)  # Create a tensor directly on the GPU
print(x_gpu)

tensor([[1., 1.],
        [1., 1.],
        [1., 1.]])


Warum sollten wir Code überhaupt auf der GPU ausführen? Weil es viel **schneller** ist. (Faktor > 100 ohne Probleme möglich)

Vergleichen wir nun die Geschwindigkeit.

In [77]:
if torch.cuda.is_available():
    size = 5000

    x = torch.randn(size, size)

    ## CPU version
    start_time = time.time()
    _ = torch.matmul(x, x)
    end_time = time.time()
    print(f"CPU time: {(end_time - start_time):6.5f}s")

    ## GPU version
    x = x.to(device)
    _ = torch.matmul(x, x)  # First operation to 'burn in' GPU
    # CUDA is asynchronous, so we need to use different timing functions
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    _ = torch.matmul(x, x)
    end.record()
    torch.cuda.synchronize()  # Waits for everything to finish running on the GPU
    print(f"GPU time: {0.001 * start.elapsed_time(end):6.5f}s")  # Milliseconds to seconds

Wir sehen also, mit *cuda* ist wirklich alles **viel** schneller.

**Hinweis:** Mit *cuda*, also auf der GPU müssen wir nun aber auch beachten, dass nun die Daten den GPU Speicher befüllen (GPU Memory). Dieser ist meistens geringer als der RAM, somit müssen die Modelle ggf. kleiner gemacht werden, bzw. das Dataset muss aufgeteilt werden (Details dazu gibt in einem anderen Notebook (Datasets und Dataloader)).

![Tensors_Everywhere](../resources/Tensors_Everywhere.jpg)

# Bonus: Google Colab

Was, wenn wir keine Grafikkarte haben und trotzdem ein Model hardwarebeschleunigt trainieren wollen?

Es gibt zahlreiche Online Anbieter, wo man gratis Notebooks laufen lassen kann inkl. GPU. Zum Beispiel [Google Colab](https://colab.research.google.com/).

Wir laden nun unser Notebook hier hoch und testen Colab ein wenig.

**Vorteile**:
* Gratis
* Gute GPU's
* Ermöglicht Verwendung auch mit wenig-leistungsfähigem Laptop

**Nachteile:**
* Runtime nur für begrenzte (inaktive) Zeit
* Hochladen von Daten nervig

## Aufgabe:

Programme ein mehrdimensionales Linear Regression Modell für 4 Input Features $x_1, x_2, x_3, x_4$ und berechne den Gradient bzgl. der Parameter $w_1, w_2, w_3, w_4, x_1, x_2, x_3, x_4, b$.

Bearbeite dafür die folgenden Schritte:
* Eingabe: Vektor $x$ der mit 4 Features (Tensor mit Shape (4))
* Gewichtsmatrix $W$ (Größe $1\times 4$)
* Offset $b$ (Größe $1$)

Berechne:
* $y = Wx+b = w_1\cdot x_1 + w_2 \cdot x_2 + w_3 \cdot x_3 + w_4 \cdot x_4 + b$
* a = $\max(y, 0)$
* $L=f(a)$ mit $f(a) = a^2$ (*Dummy*-Loss Funktion)
* Ableitungen bzgl. $W$, $x$ und $b$.


Die Gewichtsvektoren und Input kann zufällig gewählt werden.

**Bonus:** Verwende dazu das Attribut **device**, sodass immer die GPU verwendet wird, falls eine verfügbar ist.

**Bonus:** Lade nochmal das Iris Dataset und berechne den Output hier? Was muss alles geändert werden?

### Lösung:

---

Die vorige Aufgabenstellung war der Türöffner für "echte" neuronale Netze, mit welchen wir uns von nun an befassen wollen.